src/pubsub/ua_pubsub_receiver.c

```diff
@@ -0,0 +1,207 @@
+#else
+            UA_Variant_clear(&dsm->data.keyFrameData.dataSetFields[i].value);
+#endif
+        }
+    }
+    else if(dsm->header.fieldEncoding == UA_FIELDENCODING_DATAVALUE) {
+        for(UA_UInt16 i = 0; i < dsm->data.keyFrameData.fieldCount; i++) {
+#ifdef UA_ENABLE_PUBSUB_BUFMALLOC
+            UA_DataValue_init(&dsm->data.keyFrameData.dataSetFields[i]);
+#else
+            UA_DataValue_clear(&dsm->data.keyFrameData.dataSetFields[i]);
+#endif
+        }
+    }
+}
+
+UA_StatusCode
+readNetworkMessage(const UA_Logger *logger,
+                               UA_ByteString *buffer, size_t *currentPosition,
+                               UA_NetworkMessage *currentNetworkMessage,
+                               UA_ReaderGroup *readerGroup) {
+
+    UA_StatusCode rv = UA_STATUSCODE_GOOD;
+#ifdef UA_ENABLE_PUBSUB_ENCRYPTION
+    rv = verifyAndDecryptNetworkMessage(logger,
+                                        buffer,
+                                        currentPosition,
+                                        currentNetworkMessage,
+                                        readerGroup);
+    UA_CHECK_STATUS_WARN(rv, return rv, logger, UA_LOGCATEGORY_SERVER,
+                         "Subscribe failed. verify and decrypt network message failed.");
+#endif
+    rv = UA_NetworkMessage_decodeBinary(buffer, currentPosition, currentNetworkMessage);
+    UA_CHECK_STATUS_WARN(rv, return rv, logger, UA_LOGCATEGORY_SERVER,
+                         "Subscribe failed. receive network message failed.");
+
+    return UA_STATUSCODE_GOOD;
+}
+
+static
+UA_StatusCode
+receiveNetworkMessage(UA_Server *server, UA_ReaderGroup *readerGroup,
+                      UA_PubSubConnection *connection, size_t previousPosition,
+                      UA_ByteString *buffer, size_t *currentPosition) {
+    size_t paddingBytes = 0;
+    UA_NetworkMessage currentNetworkMessage;
+    memset(&currentNetworkMessage, 0, sizeof(UA_NetworkMessage));
+
+    UA_StatusCode rv = UA_STATUSCODE_GOOD;
+    rv = readNetworkMessage(&server->config.logger,
+                                        buffer,
+                                        currentPosition,
+                                        &currentNetworkMessage,
+                                        readerGroup);
+    UA_CHECK_STATUS_WARN(rv, goto error, &server->config.logger, UA_LOGCATEGORY_SERVER,
+                         "Subscribe failed. verify and decrypt network message failed.");
+
+    /* TODO: We already know the ReaderGroup at this point. Now we loose that
+    * information. There is only one place where
+    * UA_PubSubConnection_processNetworkMessage is used. */
+    rv = UA_PubSubConnection_processNetworkMessage(server, connection, &currentNetworkMessage);
+    UA_CHECK_STATUS_WARN(rv, goto error, &server->config.logger, UA_LOGCATEGORY_SERVER,
+        "Subscribe failed. process network message failed.");
+    UA_NetworkMessage_clear(&currentNetworkMessage);
```

Comment:
Just "fall through" to the cleanup case below (instead of return good).
And the _clear happens automatically.